packages/cli/src/commands/hydrogen/upgrade.ts

```diff
@@ -0,0 +1,2931 @@
+      char: 'd',
+    }),
+  };
+
+  async run(): Promise<void> {
+    const {flags} = await this.parse(Upgrade);
+    const appPath = flags.path ? path.resolve(flags.path) : process.cwd();
+    const version = flags.version;
+    const dryRun = Boolean(flags['dry-run']);
+    await runUpgrade({appPath, version, dryRun});
+  }
+}
+
+let CACHED_CHANGELOG: ChangeLog | null = null;
+
+export async function runUpgrade({
+  appPath,
+  dryRun,
+  version: targetVersion,
+}: {
+  appPath: string;
+  dryRun: boolean;
+  version?: string;
+}) {
+  await checkIsGitRepo(appPath);
+
+  await checkDirtyGitBranch(appPath);
+
+  const current = await getHydrogenVersion({appPath});
+
+  if (!current?.currentVersion) {
+    throw new Error('Failed to get current Hydrogen version');
+  }
+
+  const {currentVersion, currentDependencies} = current;
+
+  // TODO: swap when we merge to production
+  // const snapshots = await fetchChangelog();
+  const changelog = await fetchTempChangelog();
+
+  const {availableUpgrades} = getAvailableUpgrades({
+    releases: changelog.releases,
+    currentVersion,
+  });
+
+  if (!availableUpgrades?.length) {
+    const upToDateMessage = `You are on the latest Hydrogen version: ${getAbsoluteVersion(
+      currentVersion,
+    )}`;
+
+    renderSuccess({headline: upToDateMessage});
+
+    if (!process.env.SHOPIFY_UNIT_TEST) {
+      process.exit(0);
+    }
+
+    return;
+  }
+
+  // Prompt the user to select a version from the list of available upgrades
+  const selectedRelease = await getSelectedRelease({
+    currentVersion,
+    targetVersion,
+    availableUpgrades,
+  });
+
+  if (!selectedRelease) {
+    renderFatalError({
+      name: 'error',
+      type: 0,
+      message: 'No Hydrogen version selected',
+      tryMessage: `Please try again later`,
+    });
+
+    if (!process.env.SHOPIFY_UNIT_TEST) {
+      process.exit(0);
+    }
+
+    return;
+  }
+
+  // Get an aggregate list of features and fixes included in the upgrade versions range
+  const cumulativeRelease = getCummulativeRelease({
+    releases: changelog.releases,
+    currentVersion,
+    selectedRelease,
+  });
+
+  // Prompt the user to confirm the upgrade and display a list of features and fixes
+  const confirmed = await displayConfirmation({
+    appPath,
+    cumulativeRelease,
+    dryRun,
+    selectedRelease,
+    targetVersion,
+  });
+
+  // Exit if the user chose to return to the version selection prompt
+  if (!confirmed) {
+    return;
+  }
+
+  // skip the dependency upgrade step and validation if we're doing a dry run
+  if (!dryRun) {
+    await upgradeNodeModules({appPath, selectedRelease, currentDependencies});
+    await validateUpgrade({
+      appPath,
+      selectedRelease,
+    });
+  }
+
+  // Generate a markdown file with upgrade instructions
+  const instrunctionsFilePath = await generateUpgradeInstructionsFile({
+    appPath,
+    cumulativeRelease,
+    currentVersion,
+    dryRun,
+    selectedRelease,
+  });
```

Comment:
Considering `generateUpgradeInstructionsFile` has prompts, is this maybe something we could do in parallel?

```js
const updatingPromise = upgradeNodeModules().then(validateUpgrade).catch(removeInstructionsFile);
const generatingFile = generateUpgradeInstructionsFile();

await renderTask([...]);
```

But for that we need upgradeNodeModules to be independent from renderTasks.